pkg/hub/managedcluster/deletion_controller.go

```diff
@@ -0,0 +1,303 @@
+			}),
+		)
+
+		if updatedErr != nil {
+			return updatedErr
+		}
+
+		syncCtx.Queue().AddAfter(managedCluster.Name, time.Duration(estimate))
+		return nil
+	}
+
+	if remaining.numRemainingResource > 0 {
+		_, _, updatedErr := helpers.UpdateManagedClusterStatus(
+			ctx,
+			c.clusterClient,
+			managedClusterName,
+			helpers.UpdateManagedClusterConditionFn(metav1.Condition{
+				Type:    ConditionTypeDeleteSuccess,
+				Status:  metav1.ConditionFalse,
+				Reason:  ResourceRemainReason,
+				Message: fmt.Sprintf("resource %s for cluster %s has %d resource remaining", remaining.resource.String(), managedCluster.Name, remaining.numRemainingResource),
+			}),
+		)
+
+		if updatedErr != nil {
+			return updatedErr
+		}
+
+		syncCtx.Queue().AddAfter(managedCluster.Name, time.Duration(estimate))
+		return nil
+	}
+
+	return c.removeManagedClusterFinalizer(ctx, managedCluster)
+}
+
+type totalRemainingResource struct {
+	resource               schema.GroupVersionResource
+	numRemainingResource   int
+	numRemainingFinalizers map[string]int
+}
+
+func (c *managedClusterDeletionController) cleanup(ctx context.Context, managedClusterName string) (totalRemainingResource, error) {
+	// monitor predefined resource at first, do not delete anything until all resource defined here has been cleaned by other controller.
+	for _, gvr := range c.preDeleteMonitorResources {
+		remaining, err := c.monitorGVR(ctx, managedClusterName, gvr, metav1.ListOptions{})
+		if err != nil || remaining.numRemainingResource > 0 {
+			return remaining, err
+		}
+	}
+
+	// delete all managedcluster addons.
+	remainingAddon, err := c.cleanupGVR(ctx, managedClusterName, addonv1alpha1.GroupVersion.WithResource("managedclusteraddons"), metav1.ListOptions{})
+	if err != nil || remainingAddon.numRemainingResource > 0 {
+		return remainingAddon, err
+	}
+
+	// delete all manifestworks
+	remainingWorks, err := c.cleanupGVR(ctx, managedClusterName, workapiv1.GroupVersion.WithResource("manifestworks"), metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("!%s", DeletionByOtherLabelKey),
+	})
+
+	if err != nil || remainingWorks.numRemainingResource > 0 {
+		return remainingAddon, err
+	}
+
+	// monitor all manifestworks again, this is to ensure all works have been deleted.
+	remainingWorks, err = c.monitorGVR(ctx, managedClusterName, workapiv1.GroupVersion.WithResource("manifestworks"), metav1.ListOptions{})
+	if err != nil || remainingWorks.numRemainingResource > 0 {
+		return remainingAddon, err
+	}
+
+	// for namespace deletion, we only delete ns with certain name and no deleteByOther label.
+	remainingNS, err := c.cleanupGVR(ctx, metav1.NamespaceAll, corev1.SchemeGroupVersion.WithResource("namespaces"), metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("!%s", DeletionByOtherLabelKey),
+		FieldSelector: fmt.Sprintf("metadata.name=%s", managedClusterName),
+	})
```

Comment:
should we delete the rolebidings of registration/work agent if the cluster namespace has DeletionByOtherLabel?

Reply:
I think so, we do not want that permission exists when cluster is deleted.